File: src/main/java/org/lanternpowered/server/text/translation/TextTranslation.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.text.translation;

import static com.google.common.base.Preconditions.checkNotNull;

import com.google.common.base.MoreObjects;
import org.lanternpowered.server.text.LanternTextSerializer;
import org.spongepowered.api.item.inventory.Inventory;
import org.spongepowered.api.text.Text;
import org.spongepowered.api.text.serializer.TextSerializers;
import org.spongepowered.api.text.translation.Translation;

import java.util.Locale;
import java.util.Objects;

/**
 * A {@link Translation} that contains a {@link Text} object internally.
 * The {@link #toText(Translation)} methods can be used to convert
 * {@link Translation} into a {@link Text} object.
 * <p>
 * The main usage for these {@link Translation}s are within the
 * {@link Inventory}s, these use always a {@link Translation} as name.
 */
public final class TextTranslation implements Translation {

    /**
     * Constructs a new {@link TextTranslation} for the provided
     * {@link Text} object.
     *
     * @param text The text
     * @return The text translation
     */
    public static TextTranslation of(Text text) {
        return new TextTranslation(checkNotNull(text, "text"));
    }

    public static Text toText(Translation translation) {
        return translation instanceof TextTranslation ? ((TextTranslation) translation).text : Text.of(translation);
    }

    public static Text toText(Translation translation, Object... args) {
        return translation instanceof TextTranslation ? ((TextTranslation) translation).text : Text.of(translation, args);
    }

    private final Text text;

    private TextTranslation(Text text) {
        this.text = text;
    }

    @Override
    public String getId() {
        return this.text.toPlain();
    }

    @Override
    public String get(Locale locale) {
        return ((LanternTextSerializer) TextSerializers.PLAIN).serialize(this.text, locale);
    }

    @Override
    public String get(Locale locale, Object... args) {
        return get(locale);
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
                .add("text", this.text)
                .toString();
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof TextTranslation && ((TextTranslation) obj).text.equals(this.text);
    }

    @Override
    public int hashCode() {
        return Objects.hash(getClass().getName(), this.text);
    }
}

File: src/main/kotlin/org/lanternpowered/api/text/translation/TextTranslation.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.api.text.translation

import org.lanternpowered.api.text.Text
import org.lanternpowered.api.text.toPlain
import org.spongepowered.api.text.translation.locale.Locales
import java.util.Locale

/**
 * Constructs a new [ConstantTextTranslation] from the given [Text].
 */
fun textTranslationOf(text: Text): ConstantTextTranslation = ConstantTextTranslation(text)

/**
 * Represents a [Translation] that can be translated into a
 * [Text] object, which means that it can contain formatting.
 */
interface TextTranslation : Translation {

    /**
     * Gets the default translation [Text] without extra parameters.
     *
     * @return The default translation text
     */
    @JvmDefault
    fun getText(): Text = getText(Locales.DEFAULT)

    /**
     * Gets the translation text without any parameters replaced.
     *
     * @param locale The language to get the translated format string for
     * @return The translation text without any parameters
     */
    fun getText(locale: Locale): Text

    @JvmDefault
    override fun get(): String = getText().toPlain()

    @JvmDefault
    override fun get(locale: Locale): String  = getText(locale).toPlain()
}

/**
 * Represents a [TextTranslation] that holds a constant [Text] object.
 */
data class ConstantTextTranslation(private val text: Text) : TextTranslation {
    override fun get(locale: Locale): String = this.text.toPlain(locale)
    override fun getText(): Text = this.text
    override fun getText(locale: Locale): Text = this.text
    override fun getId(): String = get()
}
